File: src/main/java/dev/jsinco/luma/lumaitems/enums/ToolType.java
package dev.jsinco.luma.lumaitems.enums;

import org.bukkit.Material;
import org.bukkit.inventory.EquipmentSlot;

import java.util.List;

// Enum for generic tooltypes
public enum ToolType {
    ARMOR(List.of("HELMET", "CHESTPLATE", "LEGGINGS", "BOOTS")),
    WEAPON(List.of("SWORD", "AXE", "BOW", "CROSSBOW", "TRIDENT", "SHIELD", "MACE")),
    TOOL(List.of("PICKAXE", "AXE", "SHOVEL", "HOE", "ROD"));

    private final List<String> gearType;

    ToolType(List<String> list) {
        this.gearType = list;
    }

    public List<String> getGearTypes() {
        return gearType;
    }

    public static List<String> getArmorStrings() {
        return ARMOR.getGearTypes();
    }

    public static List<String> getWeaponStrings() {
        return WEAPON.getGearTypes();
    }

    public static List<String> getToolStrings() {
        return TOOL.getGearTypes();
    }

    public static ToolType getToolType(Material material) {
        for (String string :  ToolType.getArmorStrings()) {
            if (material.toString().contains(string)) return ToolType.ARMOR;
        }
        for (String string :  ToolType.getToolStrings()) {
            if (material.toString().contains(string)) return ToolType.TOOL;
        }
        for (String string :  ToolType.getWeaponStrings()) {
            if (material.toString().contains(string)) return ToolType.WEAPON;
        }
        return null;
    }


    public EquipmentSlot getEquipmentSlot() {
        switch (this) {
            case ARMOR -> {
                return EquipmentSlot.CHEST;
            }
            case WEAPON, TOOL -> {
                return EquipmentSlot.HAND;
            }
        }
        return null;
    }
}

File: src/main/java/dev/lumas/lumaitems/enums/ToolType.kt
package dev.lumas.lumaitems.enums

import java.util.Locale
import org.bukkit.Material
import org.bukkit.inventory.ItemStack

/**
 * Enum for generic Minecraft tool types.
 * Determine their type without regard for the type of material.
 */
enum class ToolType {
    HELMET,
    CHESTPLATE,
    LEGGINGS,
    BOOTS,
    SWORD,
    SPEAR,
    PICKAXE,
    AXE,
    SHOVEL,
    HOE,
    CROSSBOW,
    BOW,
    TRIDENT,
    SHIELD,
    ELYTRA,
    FISHING_ROD,
    MAGICAL,
    MACE,
    SHEARS;

    fun matches(material: Material): Boolean {
        return getToolType(material) == this
    }

    companion object {
        val magicMaterials = listOf("BLAZE_ROD", "BREEZE_ROD")


        fun getToolType(item: ItemStack): ToolType? {
            return getToolType(item.type.toString())
        }

        fun getToolType(material: Material): ToolType? {
            return getToolType(material.toString())
        }

        fun getToolType(string: String): ToolType? {
            var string = string
            string = string.uppercase(Locale.getDefault())
            if (magicMaterials.contains(string)) {
                return ToolType.MAGICAL
            }

            for (toolType in ToolType.entries) {
                if (string.contains(toolType.toString())) {
                    return toolType
                }
            }
            return null
        }
    }
}
